### PR TITLE
*: collect transaction write duration/throughput metrics for SLI/SLO (#23462)

### PR DESCRIPTION
**Pull request by ti-srebot**
cherry-pick #23462 to release-4.0
You can switch your code base to this Pull Request by using [git-extras](https://github.com/tj/git-extras):
```bash
# In tidb repo:
git pr https://github.com/pingcap/tidb/pull/23658
```

After apply modifications, you can push your change to this PR via:
```bash
git push git@github.com:ti-srebot/tidb.git pr/23658:release-4.0-f9708e604376
```

---

### What problem does this PR solve?

This PR add metrics for transaction duration/throughput, this is for DBaaS SLI/SLO.

- For the transaction which affects rows more than 20 rows, or the transaction total write key/value size more than 1MB,
we collect the transaction write throughput, which is calculated by:

```
(total_write_key/value_size) / (total_time_of_execution_write_statement + txn_commit_time)
```

- Otherwise, we think this is a small transaction, we only collect the transaction write duration, which is:

```sql
total_time_of_execution_write_statement + txn_commit_time
```

Currently, the following case will not collet the transaction write duration/throughput SLI:

1. The transaction contains `insert|replace into ... select ... from ...` statement.
2. The write SQL statement has more `read keys` than `write keys`.

The related Grafana panel display is:

image

image

### What is changed and how it works?

N/A

### Related changes

- Need to cherry-pick to the release branch

### Check List 

Tests 

- Unit test
- Manual test

Side effects

- Performance regression
    - Consumes more CPU
    - Consumes more MEM

### Release note 

- N/A

### CODE REVIEW

**Review by AilinKid — COMMENTED** [review]
pls, resolve some conflicts above.

**Comment by crazycs520** [comment]
Should cherry-pick https://github.com/pingcap/tidb/pull/21562 first.

**Comment by crazycs520** [comment]
merge https://github.com/pingcap/tidb/pull/24289 first

**Comment by crazycs520** [comment]
merge https://github.com/pingcap/tidb/pull/24289 first.

**Comment by XuHuaiyu** [comment]
This is a new feature, why do we need to cherry-pick this to release-4.0?

**Comment by crazycs520** [comment]
> This is a new feature, why do we need to cherry-pick this to release-4.0?

DBaaS need this to implement SLI/SLO calculation.

File: session/txn.go [hunk 1]
```diff
@@ -17,6 +17,7 @@ import (
 	"bytes"
 	"context"
 	"fmt"
+	"github.com/pingcap/tidb/util/sli"
```
Comment (tangenta) [inline session/txn.go:20]: Please group this import properly.

**Pull request merged**